net/libfab/libfab.c

```diff
@@ -0,0 +1,2963 @@
+{
+	m0_mutex_lock(&tm->ftm_ntm->ntm_mutex);
+}
+
+/**
+ * Used to unlock the transfer machine mutex
+ */
+static inline void libfab_tm_unlock(struct m0_fab__tm *tm)
+{
+	m0_mutex_unlock(&tm->ftm_ntm->ntm_mutex);
+}
+
+static inline int libfab_tm_trylock(struct m0_fab__tm *tm)
+{
+	return m0_mutex_trylock(&tm->ftm_ntm->ntm_mutex);
+}
+
+/**
+ * Used to lock the transfer machine event post mutex
+ */
+static inline void libfab_tm_evpost_lock(struct m0_fab__tm *tm)
+{
+	m0_mutex_lock(&tm->ftm_evpost);
+}
+
+/**
+ * Used to unlock the transfer machine event post mutex
+ */
+static inline void libfab_tm_evpost_unlock(struct m0_fab__tm *tm)
+{
+	m0_mutex_unlock(&tm->ftm_evpost);
+}
+
+/**
+ * Used to check if the transfer machine mutex is locked.
+ * Returns true if locked, else false.
+ */
+static inline bool libfab_tm_is_locked(const struct m0_fab__tm *tm)
+{
+	return m0_mutex_is_locked(&tm->ftm_ntm->ntm_mutex);
+}
+
+/**
+ * Helper function that posts a tm state change event.
+ */
+static void libfab_tm_event_post(struct m0_fab__tm *tm,
+				 enum m0_net_tm_state state)
+{
+	struct m0_net_end_point *listen = NULL;
+
+	if (state == M0_NET_TM_STARTED) {
+		/* Check for LISTENING Passive endpoint */
+		listen = &tm->ftm_pep->fep_nep;
+		M0_ASSERT(listen != NULL);
+	}
+	
+	m0_net_tm_event_post(&(struct m0_net_tm_event) {
+		.nte_type       = M0_NET_TEV_STATE_CHANGE,
+		.nte_next_state = state,
+		.nte_time       = m0_time_now(),
+		.nte_ep         = listen,
+		.nte_tm         = tm->ftm_ntm,
+	});
+}
+
+/**
+ * Finds queued buffers that timed out and completes them with a
+ * ETIMEDOUT error.
+ */
+static void libfab_tm_buf_timeout(struct m0_fab__tm *ftm)
+{
+	struct m0_net_transfer_mc *net = ftm->ftm_ntm;
+	int                        i;
+	m0_time_t                  now = m0_time_now();
+
+	M0_PRE(libfab_tm_is_locked(ftm));
+	M0_PRE(libfab_tm_invariant(ftm));
+
+	for (i = 0; i < ARRAY_SIZE(net->ntm_q); ++i) {
+		struct m0_net_buffer *nb;
+
+		m0_tl_for(m0_net_tm, &net->ntm_q[i], nb) {
+			if (nb->nb_timeout < now) {
+				nb->nb_flags |= M0_NET_BUF_TIMED_OUT;
+				libfab_buf_done(nb->nb_xprt_private,
+						-ETIMEDOUT);
+			}
+		} m0_tl_endfor;
+	}
+	M0_POST(libfab_tm_invariant(ftm));
+}
+
+/**
+ * Finds pending buffers completions and completes them.
+ *
+ * A buffer is placed on tm::ftm_done queue when its operation is done, but the
+ * completion call-back cannot be immediately invoked, for example, because
+ * completion happened in a synchronous context.
+ */
+static void libfab_tm_buf_done(struct m0_fab__tm *ftm)
+{
+	struct m0_fab__buf *buffer;
+	int                 nr = 0;
+
+	M0_PRE(libfab_tm_is_locked(ftm) && libfab_tm_invariant(ftm));
+	m0_tl_for(fab_buf, &ftm->ftm_done, buffer) {
+		fab_buf_tlist_del(buffer);
+		libfab_buf_complete(buffer, buffer->fb_status);
+		nr++;
+	} m0_tl_endfor;
+
+	if (nr > 0 && ftm->ftm_ntm->ntm_callback_counter == 0)
+		m0_chan_broadcast(&ftm->ftm_ntm->ntm_chan);
+	M0_POST(libfab_tm_invariant(ftm));
```

Comment:
Same as above, do we need libfab_tm_invariant in both place (when we enter the function line 523 and when we leave) ?

Reply:
This is a sanity check on the transfer machine parameters after we have completed processing the ftm_done list.